File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LEN 200 + 1
#define MAX_HORSE_MOVES 30 + 1

typedef enum
{
    false,
    true
} bool;
typedef struct _node
{
    int y;
    int x;
    int moves;
    int horse_moves;
    struct _node *next;
} node;

typedef struct _queue
{
    node *front;
    node *rear;
    int count;
} queue;

int HORSE_VEC[][2] = {{-2, -1}, {-1, -2}, {1, -2}, {2, -1}, {2, 1}, {1, 2}, {-1, 2}, {-2, 1}};
int VEC[][2] = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

int k, w, h;
int space[MAX_LEN][MAX_LEN];
bool visited[MAX_HORSE_MOVES][MAX_LEN][MAX_LEN];

node *init_node(int y, int x, int moves, int horse_moves)
{
    node *n = (node *)malloc(sizeof(node));
    n->y = y;
    n->x = x;
    n->moves = moves;
    n->horse_moves = horse_moves;

    return n;
}
void init_queue(queue *q)
{
    q->front = NULL;
    q->rear = NULL;
    q->count = 0;
}
void enque(queue *q, node *n)
{
    if (!q->count)
    {
        q->front = n;
        q->rear = q->front;
        q->front->next = NULL;
    }
    else
    {
        q->rear->next = n;
        q->rear = q->rear->next;
    }
    q->count++;
}

node deque(queue *q)
{
    node n;

    memcpy(&n, q->front, sizeof(node));

    node *tmp = q->front;
    q->front = q->front->next;
    q->count--;
    free(tmp);
    return n;
}

void traverse(queue *q, int vec[][2], int vec_len, node *n, int offset)
{
    for (int i = 0; i < vec_len; i++)
    {
        int ny, nx;
        ny = n->y + vec[i][0];
        nx = n->x + vec[i][1];
        if (0 <= ny && ny < h && 0 <= nx && nx < w && space[ny][nx] != 1 && !visited[n->horse_moves + offset][ny][nx])
        {
            visited[n->horse_moves + offset][ny][nx] = true;
            enque(q, init_node(ny, nx, n->moves + 1, n->horse_moves + offset));
        }
    }
}
int main()
{
    queue *q = (queue *)malloc(sizeof(queue));
    init_queue(q);
    node *n = init_node(0, 0, 0, 0);
    enque(q, n);

    scanf("%d", &k);
    scanf("%d %d", &w, &h);
    for (int i = 0; i < h; i++)
    {
        for (int j = 0; j < w; j++)
        {
            scanf(" %d", &space[i][j]);
        }
    }

    while (q->count)
    {
        node front_node = deque(q);
        if (front_node.x == w - 1 && front_node.y == h - 1)
        {
            printf("%d\n", front_node.moves);
            exit(0);
        }
        if (front_node.horse_moves < k)
        {
            traverse(q, HORSE_VEC, 8, &front_node, 1);
        }
        traverse(q, VEC, 4, &front_node, 0);
    }

    printf("-1\n");
    return 0;
}

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
from typing import *
from collections import deque

HORSE_VEC = ((-2, -1), (-1, -2), (1, -2), (2, -1), (2, 1), (1, 2), (-1, 2), (-2, 1))
VEC = ((1, 0), (-1, 0), (0, 1), (0, -1))

if __name__ == '__main__':
    k = int(input())
    w, h = list(map(int, input().split()))
    space: List[List[int]] = []
    visited: List[List[List[bool]]] = [[[False] * w for _ in range(h)] for _ in range(k + 1)]
    Q: Deque[Tuple[int, int, int, int]] = deque()
    Q.append((0, 0, 0, 0))

    for _ in range(h):
        space.append(list(map(int, input().split())))

    while Q:
        y, x, moves, horse_moves = Q.popleft()
        if y == h - 1 and x == w - 1:
            print(moves)
            exit(0)

        if horse_moves < k:
            for dy, dx in HORSE_VEC:
                ny, nx = y + dy, x + dx
                if 0 <= ny < h and 0 <= nx < w and space[ny][nx] != 1 and not visited[horse_moves + 1][ny][nx]:
                    visited[horse_moves + 1][ny][nx] = True
                    Q.append((ny, nx, moves + 1, horse_moves + 1))

        for dy, dx in VEC:
            ny, nx = y + dy, x + dx
            if 0 <= ny < h and 0 <= nx < w and space[ny][nx] != 1 and not visited[horse_moves][ny][nx]:
                visited[horse_moves][ny][nx] = True
                Q.append((ny, nx, moves + 1, horse_moves))

    print(-1)
